lib/pub_sub.js

```diff
@@ -18,34 +17,40 @@ class PubSub {
       pub: new RedisClient(),
       sub: new RedisClient()
     };
-    this.redis.sub.on('message', this.emitMessage);
-    this.redis.sub.on('pmessage', this.emitMessage);
+    this.subscribers = {};
   }
 
   emitMessage(pattern, channel, message) {
     return this.emitter.emit(pattern, JSON.parse(message || channel));
   }
 
-  subscribe(pattern, fn) {
-    var method;
-    method = pattern.match(/\*|\[|\]|\?/) ? 'psubscribe' : 'subscribe';
-    return this.redis.sub[`${method}Async`](pattern).then(() => {
-      return this.emitter.on(pattern, fn);
+  async subscribe(pattern, fn) {
+    let subscriber = this.subscribers[pattern];
+    if (!subscriber) {
+      subscriber = this.redis.sub.duplicate();
+      await subscriber.connect();
+      this.subscribers[pattern] = subscriber;
+    }
```

Comment:
Tests fail unless I duplicate the subscription client here and create a dedicated connection for each channel.

Reply:
All the methods are asynchronous now. In v3, we wrap them in `promisify` and call `punsubscribeAsync`. 

I’m confused though: Redis has always used promises in Sugar. We’ve never used the synchronous versions of the Redis commands, and they’re completely removed in v4.

Reply:
The change to dedicated connections is straight from the v4 Readme. 
https://github.com/redis/node-redis/blob/master/docs/pub-sub.md

Reply:
hmm - that appears to use async to await the connection then proceed to use that connection. However we now call `client.connect();` and then proceed to use that connection in tests, https://github.com/zooniverse/sugar/blob/dc62ddfbbddb0390ad1eeb01388db9e15088af0f/lib/redis_client.js#L26

I may be completely wrong but the connection did seem to have issues for those failing tests and your change to create new client and await the connection seems to fix them. So perhaps we're not setting it up right? 

Does adding the following client event event info logs to the redis client constructor help? 
```
client.on('connect', () => console.log('connect'));
client.on('ready', () => console.log('ready'));
client.on('reconnecting', () => console.log('reconnecting'));
client.on('error', (error) => console.log(error));
client.on('end', () => console.log('end'));
```

Reply:
>We don't really need to know this function correctly resolved before we setup the emitter. I think this can move back to returning a promise and remove the async setup.

We could do that, but then we’re  setting up the emitter regardless of whether the subscription succeeded or failed, if I’m reading the PubSub changes correctly. 

>The subscribe-like commands return a promise. If the command is executed successfully the promise will be fulfilled, otherwise the promise will be rejected.

https://github.com/redis/node-redis/blob/master/docs/v3-to-v4.md#no-message-like-event

Reply:
This function is asynchronous, by the way, so the implementation hasn’t changed there, just the syntax. It still returns a Promise.

Reply:
It currently works async via the promise fulfillment, https://github.com/zooniverse/sugar/blob/b8fd1c7d16ca9a13db44626a323f6ee3d01f071f/lib/pub_sub.js#L32-L34

yes - however in your new version we await the result before configuring the emitter which broke the tests when i looked at it. 

Setting up a new redis connection for the different channels will consumer a lot of resources and hit the redis server hard. The current node-redis system looks like it uses a connection pool, so there shouldn't be a need for creating these new connection. 

i think the new client changes are related to test harness failures as it doesn't appear to impact server / app behaviours.

Reply:
>yes - however in your new version we await the result before configuring the emitter which broke the tests when i looked at it.

The original code awaited the Promise too, then configured the emitter when it resolves. I changed it for readability, but didn’t alter what it’s doing.

Reply:
The v4 docs specifically say that PubSub must use `client.duplicate()` now but I don’t know. 🤷🏻‍♂️

Reply:
You can check that by removing the `.duplicate()`, at which point the new subscription never resolves.

Reply:
fine - when i was testing the promise never fulfilled - instead it just hung till it hit a timeout. 

I did not see those docs at all, well there you go - https://github.com/redis/node-redis/blob/00e365250ecb45990118739e624a64c3b1183446/docs/pub-sub.md?plain=1#L7-L12

Ignore all of my comments above, this one should be relevant https://github.com/zooniverse/sugar/pull/129#issuecomment-1404688215 but if the lib client demands new connections then give go for it.

Reply:
That will be the same regardless of the Promise syntax that you use: `.then()` or `await`.

Reply:
Just to be clear, these two functions both return the same Promise, so reverting back to `.then()` alone won't fix the issue. The `async/await` syntax is just there to make the original nested code easier to read. So I wouldn't spend a lot of time trying to fix this with `.then()`. I think that's more about which JS syntax you find easier to follow.

```js
function subscribe() {
  const client = redis.createClient();

  return client.connect().then(function () {
   return client.subscribe(channel, listener).then(function () {
     emitter.on('message', fn);
   });
  });
}
```

```js
async function subscribe() {
  const client = redis.createClient();

  await client.connect();
  await client.subscribe(channel, listener);
  return emitter.on('message', fn);
}
```

Both those functions return a pending Promise, which will resolve if the client successfully connects and subscribes to a channel.

Reply:
>fine - when i was testing the promise never fulfilled - instead it just hung till it hit a timeout.

Yes! That's exactly the same problem I've been running into, both yesterday and today. If I add `client.on('connect', () => console.log('connected!')` to the `RedisClient` class, then I see the client connect but the `client.subscribe()` command never resolves, unless I call `client.duplicate()` before subscribing.

Here's the terminal output. All the PubSub tests time out after 2s and fail.

<img width="610" alt="Screenshot of the PubSub tests in a terminal. The client is connecting, but tests are timing out and failing." src="https://user-images.githubusercontent.com/59547/214802808-8858db9e-f773-4127-b04a-6d0d99fb2fc5.png">